### PR TITLE
remove Default from CandidateDescriptor

### PR DESCRIPTION
**Pull request by drahnr**
Introduces a set of helpers for testing to create `Candidate*`-types with ease. The purpose of these types twofold, 1) subsystem unit tests which are in need of unified fillers, that create types which can have arbitrary content 2) for benchmarking where the content has influence on the measured runtime.

@emostov please take over and drive this to completion ASAP

TODOs

* [x] remove remainaing `fn dummy() -> Self` implementations @emostov 
* [x] introduced a test failure  ``` `sr25519_sign_version_1` called outside of an Externalities-provided environment.' ```, needs a fix
* [x] move relevant `fn`s into the the test helper crate (i.e. `CandidateBuilder` is a good candidate, there are 4 or 5 copies of that. Also `SessionInfo`) @drahnr , can be defered to a follow up PR
* [x] verify the runtime benchmarks covers the input parameter space sufficiently, no `dummy_*` calls in there @emostov 
* [x] Avoid `x: Default::default()` usage as much as possible @drahnr
* [x] format

Ref paritytech/substrate#10403

skip check-dependent-cumulus

### CODE REVIEW

File: node/core/candidate-validation/src/tests.rs [hunk 1]
```diff
@@ -25,18 +25,45 @@ use polkadot_primitives::v1::{HeadData, UpwardMessage};
 use sp_core::testing::TaskExecutor;
 use sp_keyring::Sr25519Keyring;
 
-fn collator_sign(descriptor: &mut CandidateDescriptor, collator: Sr25519Keyring) {
-	descriptor.collator = collator.public().into();
+
+fn dummy_hash() -> Hash {
+	Hash::repeat_byte(0u8)
```
Comment (ordian) [inline node/core/candidate-validation/src/tests.rs:30]:
```suggestion
	Hash::zero()
```

File: node/core/candidate-validation/src/tests.rs [hunk 2]
```diff
@@ -25,18 +25,45 @@ use polkadot_primitives::v1::{HeadData, UpwardMessage};
...
-		&descriptor.persisted_validation_data_hash,
-		&descriptor.pov_hash,
-		&descriptor.validation_code_hash,
+		&relay_parent,
+		&para_id,
+		&persisted_validation_data_hash,
+		&pov_hash,
+		&validation_code_hash,
 	);
+	let zeros = Hash::repeat_byte(0u8);
```
Comment (ordian) [inline node/core/candidate-validation/src/tests.rs:50]:
```suggestion
	let zeros = dummy_hash();
```

File: primitives/test-helpers/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+		new_validation_code: None,
+		horizontal_messages: vec![],
+		processed_downward_messages: 0,
+		hrmp_watermark: 0_u32,
+	}
+}
+
+/// Create meaningless dummy hash.
+pub fn dummy_hash() -> Hash {
+	// TODO make this PRNG based
```
Comment (emostov) [inline primitives/test-helpers/src/lib.rs:79]: @drahnr are you ok with leaving this out for now and leaving for a follow up?

Comment (emostov) [reply primitives/test-helpers/src/lib.rs:82]: It would make things a little more complicated because we would no longer be able to use this when we are asserting the output

File: node/core/dispute-coordinator/src/lib.rs [hunk 4]
```diff
@@ -25,9 +25,6 @@
 //! another node, this will trigger the dispute participation subsystem to recover and validate the block and call
 //! back to this subsystem.
 
-#![forbid(unused_must_use)]
-#![forbid(unused_crate_dependencies)]
-
```
Comment (emostov) [inline node/core/dispute-coordinator/src/lib.rs]: @drahnr Is this ok? Otherwise we would need to feature gate most of the deps in this crate, which imo should be done in another PR

File: runtime/kusama/src/weights/runtime_parachains_paras_inherent.rs [hunk 5]
```diff
@@ -146,10 +145,10 @@ impl<T: frame_system::Config> runtime_parachains::paras_inherent::WeightInfo for
 	// Storage: Hrmp HrmpWatermarks (r:0 w:1)
 	// Storage: Paras Heads (r:0 w:1)
 	fn enter_backed_candidates_variable(v: u32, ) -> Weight {
-		(319_740_000 as Weight)
-			// Standard Error: 26_000
-			.saturating_add((49_051_000 as Weight).saturating_mul(v as Weight))
-			.saturating_add(T::DbWeight::get().reads(29 as Weight))
+		(1_134_397_000 as Weight)
```
Comment (emostov) [inline runtime/kusama/src/weights/runtime_parachains_paras_inherent.rs:148]: 3x increase to the static weight here is quite significant

**Comment by emostov** [comment]
/benchmark runtime polkadot runtime_parachains::paras_inherent

File: node/core/dispute-coordinator/src/lib.rs (hunk 4 above)
Comment (drahnr) [reply node/core/dispute-coordinator/src/lib.rs]: I just introduced it as part of this PR, so it's alright, removed two deps :)

File: runtime/kusama/src/weights/runtime_parachains_paras_inherent.rs (hunk 5 above)
Comment (drahnr) [reply runtime/kusama/src/weights/runtime_parachains_paras_inherent.rs:148]: I don't quite understand where that comes from

File: primitives/test-helpers/src/lib.rs (hunk 3 above)
Comment (drahnr) [reply primitives/test-helpers/src/lib.rs:82]: Please remove the TODOs, it's fine for right now, and I think we should introduce a separate fn with `_rand` suffix

File: node/network/availability-distribution/src/tests/mock.rs [hunk 6]
```diff
@@ -114,7 +116,11 @@ impl TestCandidateBuilder {
 				pov_hash: self.pov_hash,
 				relay_parent: self.relay_parent,
 				erasure_root: self.erasure_root,
-				..Default::default()
+				collator: CollatorId::from(sr25519::Public::from_raw([42; 32])),
+				persisted_validation_data_hash: dummy_hash(),
+				signature: sr25519::Signature::from_raw([42; 64]).into(),
```
Comment (gavofyork) [inline node/network/availability-distribution/src/tests/mock.rs:121]: Doesn't get its own helper?

File: runtime/parachains/src/builder.rs [hunk 7]
```diff
@@ -47,7 +47,7 @@ fn dummy_validation_code() -> ValidationCode {
 /// "features = runtime-benchmarks".
 fn account<AccountId: Decode + Default>(name: &'static str, index: u32, seed: u32) -> AccountId {
 	let entropy = (name, index, seed).using_encoded(sp_io::hashing::blake2_256);
-	AccountId::decode(&mut &entropy[..]).unwrap_or_default()
+	AccountId::decode(&mut &entropy[..]).expect("256 bit input is valid. qed.")
```
Comment (gavofyork) [inline runtime/parachains/src/builder.rs:50]: This is fine for now, but in general it's best to use `TrailZeroInput` to ensure that `AccountId::decode` doesn't run out of input data. In the future, we should just have a trait function for `fn from_entropy(e: impl Encode) -> Self`.

File: runtime/polkadot/src/weights/runtime_parachains_paras_inherent.rs [hunk 8]
```diff
@@ -69,9 +69,9 @@ impl<T: frame_system::Config> runtime_parachains::paras_inherent::WeightInfo for
 	// Storage: Hrmp HrmpWatermarks (r:0 w:1)
 	// Storage: Paras Heads (r:0 w:1)
 	fn enter_variable_disputes(v: u32, ) -> Weight {
-		(201_142_000 as Weight)
-			// Standard Error: 2_000
-			.saturating_add((304_000 as Weight).saturating_mul(v as Weight))
+		(404_051_000 as Weight)
```
Comment (gavofyork) [inline runtime/polkadot/src/weights/runtime_parachains_paras_inherent.rs:72]: ouch - 2x increase here :(

**Review by gavofyork — APPROVED** [review]

File: node/network/availability-distribution/src/tests/mock.rs (hunk 6 above)
Comment (drahnr) [reply node/network/availability-distribution/src/tests/mock.rs:121]: A residue, `dummy_collator_signature()` exists and should be used, ty!

File: runtime/polkadot/src/weights/runtime_parachains_paras_inherent.rs (hunk 8 above)
Comment (drahnr) [reply runtime/polkadot/src/weights/runtime_parachains_paras_inherent.rs:72]: It's unclear to what caused this, or if other changes on `master` caused the change, to be checked.

File: node/core/dispute-coordinator/src/real/ordering/tests.rs [hunk 9]
```diff
@@ -142,24 +145,39 @@ fn get_block_number_hash(n: BlockNumber) -> Hash {
 	BlakeTwo256::hash(&n.encode())
 }
 
+fn make_candidate_receipt(_relay_parent: Hash) -> CandidateReceipt {
+	let zeros = dummy_hash();
+	let descriptor = CandidateDescriptor {
+		para_id: 0.into(),
+		relay_parent: get_block_number_hash(2),
```
Comment (eskimor) [inline node/core/dispute-coordinator/src/real/ordering/tests.rs:152]: Why do we pass in a relay parent, but then make one up?

**Review by eskimor — COMMENTED** [review]
Overall, looks like a great improvement!

File: node/core/dispute-coordinator/src/real/ordering/tests.rs (hunk 9 above)
Comment (drahnr) [reply node/core/dispute-coordinator/src/real/ordering/tests.rs:152]: Uh, that's a very good point... I was just very happy that this test finally passed again and focused on making the rest work, good catch :+1:

**Review by eskimor — APPROVED** [review]

**Comment by drahnr** [comment]
bot merge

**Comment by drahnr** [comment]
bot merge
